Add ctest execution to coverage reporting

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit updates the coverage CI job to also include execution of the C API tests. Right now the cext crate is showing 0 coverage because we only test it without instrumention and don't collect any coverage numbers for the C API testing. This commit adds a ctest execution to the coverage job to provide coverage numbers for our C API testing.

### Details and comments

mtreinish committed:
Add ctest execution to coverage reporting

This commit updates the coverage CI job to also include execution of the
C API tests. Right now the cext crate is showing 0 coverage because we
only test it without instrumention and don't collect any coverage
numbers for the C API testing. This commit adds a ctest execution to the
coverage job to provide coverage numbers for our C API testing.

mtreinish committed:
Install cbindgen in coverage job

jakelishman commented on .github/workflows/coverage.yml:
```diff
@@ -54,16 +55,24 @@ jobs:
+          CARGO_INCREMENTAL: 0
+          RUSTFLAGS: "-Cinstrument-coverage"
+          LLVM_PROFILE_FILE: "qiskit-%p-%m.profraw"
+
```
Might be clearer to separate out the `grcov` into its own separate step now, and make sure the `name` fields on the `unittest` and `ctest` jobs both make sense - it's starting to get a bit muddled, and a bit magic that `grcov` runs in one of two test steps.

jakelishman reviewed (commented):
The Coveralls report (clicking through to it) isn't showing an increase in the `cext` crate coverage - is this working the way you expect yet?

mtreinish commented:
> The Coveralls report (clicking through to it) isn't showing an increase in the `cext` crate coverage - is this working the way you expect yet?

Not yet, the makefile isn't setup to handle the environment variables I'm setting to enable coverage collection so it's not building the instrumented library. I was debating between either running the tests manually without make or editing the makefile at the end of day yesterday.

mtreinish committed:
Run ctests manually

mtreinish committed:
Copy coverage data from ctests to grcov working dir

mtreinish committed:
Fix source path for profile data copy

mtreinish committed:
Don't build tests in release mode

mtreinish commented:
The most recent commit has the coverage collection working now: https://coveralls.io/jobs/162457981/source_files/1455686271

However the tradeoff here is I've made a bit of a mess of the cmake file and make file to support this. Basically for coverage we want to be building in debug mode but all the C test build automation is around reusing the makefile. The makefile is setup to support building a proper C dynlib for end user consumption in release mode which make sense if you do `make clib`. The underlying issue is for coverage, and I'd argue testing more broadly we want to run in debug mode which means we can't rely on `make clib`. To test that this was the last missing piece I temporarily added a debug build to `make clib`  in addition to the release mode build, but obviously this isn't ideal. So to move forward here we need to figure out the organizational structure we want in the makefile around building the c library for testing vs end user builds to better encapsulate this difference.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into ctest-coverage

mtreinish committed:
Add split path for debug vs release mode in Makefile

This commit updates the makefile organization to add a debug mode path
and a release mode path. The debug mode path is needed for running the
ctests, while release mode is what we want to use for building the clib
for direct use. With this the makefile is able to build in either
release mode or debug mode and the targets are setup so we default to
release mode unless debug is explicitly needed.

mtreinish commented:
This should be good to go now, I split out the build targets in the makefile so we have separate paths for debug and release, and then updated the `ctest` target to use debug mode.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into ctest-coverage

mtreinish commented on CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@ enable_testing ()
 # The C API is provided by the qiskit_cext Rust crate. Its header gets
 # generated by cbindgen (see `make cheader`). It is important that this done as
 # a Release-build for cbindgen to work properly.
```
```suggestion
# generated by cbindgen (see `make cheader`).
```

mtreinish committed:
Update CMakeLists.txt

mtreinish committed:
Merge branch 'main' into ctest-coverage

Cryoris commented on .github/workflows/coverage.yml:
```diff
@@ -54,16 +55,29 @@ jobs:
+      - name: Generate ctest coverage report
+        run: |
+          cargo rustc --crate-type cdylib -p qiskit-cext -- -Cinstrument-coverage -Cincremental=false
+          cbindgen --crate qiskit-cext --output dist/c/include/qiskit.h --lang C
```
We should no longer need that 🙂 
```suggestion
```

Cryoris commented on .github/workflows/coverage.yml:
```diff
@@ -33,6 +33,7 @@ jobs:
         run: |
             cargo install grcov
             sudo apt-get install lcov
+            cargo install cbindgen
```
... and that 🙂 
```suggestion
```
I assume the other cbindgen removals should come when you update from `main`

Cryoris commented on Makefile:
```diff
@@ -128,15 +134,15 @@ $(C_DIR_INCLUDE):
 
 # Use ctest to run C API tests
-ctest: $(C_LIB_CARGO_PATH) $(C_QISKIT_H) 
+ctest: $(C_LIB_DEBUG_CARGO_PATH) $(C_QISKIT_H)
```
The tests are very small right now so compiling in Debug is fine, but wouldn't it generally make more sense to run in release mode for more speed? Or do you want to avoid building twice, once for coverage and once for the tests?

mtreinish replied:
I don't think it's really going to be an issue for a long time. Right now the compilation of the rust code and the test binary takes slightly more than 1.5min and the tests only take 0.02 sec. So we're a long time away from where I think we need to be worried about the test execution time. I generally think having the debug symbols for tests is a good thing too, although maybe not in CI.

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Merge branch 'main' into ctest-coverage

Cryoris commented on Makefile:
```diff
@@ -114,11 +116,15 @@ cformat:
 
+$(C_QISKIT_H):
+	cbindgen --crate qiskit-cext --output $(C_DIR_INCLUDE)/qiskit.h --lang C
+
```
This target already exists below, which could be why the coverage job fails. We shouldn't need to call `cbindgen` as command line tool at all anymore 🙂 I didn't try this locally but one solution could be to have something like
```make
target/qiskit.h: $(C_LIB_CARGO_PATH)   # build in release mode per default

$(C_QISKIT_H): $(C_DIR_INCLUDE) target/qiskit.h  # depend on target/qiskit.h explicitly

# the first should built target/qiskit.h as artifact, so C_QISKIT_H shouldn't retrigger the build
ctest:  $(C_LIB_DEBUG_CARGO_PATH) $(C_QISKIT_H)  
```
alternatively, we could maybe just be more explicit for the `ctest` target and explicitly move all the files in the right location, that might be easier.

mtreinish replied:
The coverage job fails because by removing the cbindgen call the header wasn't getting copied to the expected directory anymore. Since the coverage job doesn't use the makefile so we can set custom flags for compiling with instrumentation just building the library isn't enough since we need the extra copy step. I fixed that in: https://github.com/Qiskit/qiskit/pull/14046/commits/333d5d3ebad0b6758fd10e8dcef18902c5900fc6

mtreinish committed:
Merge remote-tracking branch 'origin/main' into ctest-coverage

mtreinish committed:
Copy header to expected dir

mtreinish committed:
Merge branch 'main' into ctest-coverage

Cryoris commented on Makefile:
```diff
@@ -128,15 +134,15 @@ $(C_DIR_INCLUDE):
 $(C_LIBQISKIT): $(C_DIR_LIB)  $(C_LIB_CARGO_PATH)
 	cp $(C_LIB_CARGO_PATH) $(C_DIR_LIB)/$(subst _cext,,$(C_LIB_CARGO_FILENAME))
```
```suggestion
$(C_LIBQISKIT): $(C_DIR_LIB)  $(C_LIB_RELEASE_CARGO_PATH)
	cp $(C_LIB_RELEASE_CARGO_PATH) $(C_DIR_LIB)/$(subst _cext,,$(C_LIB_CARGO_FILENAME))
```

Cryoris commented on Makefile:
```diff
@@ -114,11 +116,15 @@ cformat:
 	cargo rustc --release --crate-type cdylib -p qiskit-cext
 
+$(C_QISKIT_H):
+	cbindgen --crate qiskit-cext --output $(C_DIR_INCLUDE)/qiskit.h --lang C
```
We probably shouldn't have duplicated targets in the Makefile, running this locally also gives some warnings. I'm not quite sure how to cleanly resolve this, given that both in the debug and release mode the header is called `target/qiskit.h`. Maybe the simple solution would be to just add the full build into the `ctest` target and rely on rust not redundantly recompiling?

mtreinish replied:
I think the simplest solution is actually to split out the cheader job to just call cbindgen manually. ctest depends on compilation already with `C_LIB_DEBUG_CARGO_PATH` so we get the header created from that, so depending on the target that relies on `cp` works fine. I pushed up the change to do this: https://github.com/Qiskit/qiskit/pull/14046/commits/a8c465f64d80ca701dac7458c7446cc515d8c45d

mtreinish committed:
Update Makefile

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Deduplicate header target in makefile

mtreinish committed:
Merge branch 'main' into ctest-coverage

mtreinish committed:
Fix header generation for docs

Cryoris commented on Makefile:
```diff
@@ -125,18 +128,23 @@ $(C_DIR_LIB):
+# build the original header.
+#
+# Do not depend on this in a target without one of those as dependencies first
+$(C_QISKIT_H): $(C_DIR_INCLUDE) $(C_LIB_CARGO_PATH)
```
`C_LIB_CARGO_PATH` does not exist 🙂 Or did you want to put another target here?
```suggestion
$(C_QISKIT_H): $(C_DIR_INCLUDE)
```

Cryoris commented on tox.ini:
```diff
@@ -103,13 +103,16 @@ basepython = python3
 allowlist_externals =
   make
```
`make` would no longer be needed here, right?

mtreinish replied:
With the recent changes I didn't need to change the tox.ini anymore so this can be reverted.

Cryoris reviewed (commented):
I'm still a bit skeptical about the `Makefile` setup since we'd now re-build the header every time `make cheader` is run (it's always a "dirty" target) and rely on `cbindgen` as command line tool.

Maybe we should just optimize this for the release mode as standard path and make the `ctest` target build the library in debug mode and move it into the right location. But if you prefer this solution as is, I'm also fine 🙂

mtreinish committed:
Update Makefile

Co-authored-by: Julien Gacon <gaconju@gmail.com>

mtreinish committed:
Build in debug mode explicitly for ctest makefile target

This commit reworks the makefile to minimize the changes. It changes the
ctest target to explicitly call cargo to build in debug mode and copy
the header explicitly rather than trying to split the internal logic
based on debug mode versus release mode.

mtreinish committed:
Merge remote-tracking branch 'origin/main' into ctest-coverage

mtreinish committed:
Revert tox.ini change

mtreinish commented:
I've reworked the makefile as suggested and just am explicitly building in debug mode and not reusing the other targets. I actually like this better because it explicitly recompiles before running tests which seems nice.

Cryoris reviewed (approved):
LGTM this is a really nice addition!
